homeassistant/components/zha/api.py

```diff
@@ -0,0 +1,419 @@
+"""
+Web socket API for Zigbee Home Automation devices.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/zha/
+"""
+
+import logging
+import voluptuous as vol
+
+from homeassistant.components import websocket_api
+from homeassistant.components.zha.entities import ZhaDeviceEntity
+from homeassistant.const import ATTR_ENTITY_ID
+import homeassistant.helpers.config_validation as cv
+from .const import (
+    DOMAIN, ATTR_CLUSTER_ID, ATTR_CLUSTER_TYPE, ATTR_ATTRIBUTE, ATTR_VALUE,
+    ATTR_MANUFACTURER, ATTR_COMMAND, ATTR_COMMAND_TYPE, ATTR_ARGS, IN, OUT,
+    CLIENT_COMMANDS, SERVER_COMMANDS, SERVER)
+
+_LOGGER = logging.getLogger(__name__)
+
+TYPE = 'type'
+CLIENT = 'client'
+ID = 'id'
+NAME = 'name'
+RESPONSE = 'response'
+DEVICE_INFO = 'device_info'
+
+ATTR_DURATION = 'duration'
+ATTR_IEEE_ADDRESS = 'ieee_address'
+ATTR_IEEE = 'ieee'
+
+SERVICE_PERMIT = 'permit'
+SERVICE_REMOVE = 'remove'
+SERVICE_RECONFIGURE_DEVICE = 'reconfigure_device'
+SERVICE_SET_ZIGBEE_CLUSTER_ATTRIBUTE = 'set_zigbee_cluster_attribute'
+SERVICE_ISSUE_ZIGBEE_CLUSTER_COMMAND = 'issue_zigbee_cluster_command'
+ZIGBEE_CLUSTER_SERVICE = 'zigbee_cluster_service'
+IEEE_SERVICE = 'ieee_based_service'
+
+SERVICE_SCHEMAS = {
```

Comment:
I wonder how many of these things should really be services or should all be moved to websocket commands?

Services should be for control, not configuration. Do you think a user ever would want to create an automation or a script to cluster things?

Reply:
Cluster is a zigbee object that contains executable commands and readable / editable attributes on a device. This isn’t clustering in the sense of adding things to a cluster like server clustering for instance. Does that make sense?

Reply:
I left the attribute interaction a service because I can make a case for an automation wanting to change something like sensitivity of a device after events occur for a period of time (occupancy sensors, vibration sensors, light temperature etc and in some cases sensors / binary sensors there aren't domain services that would cover this). 

I personally have vibration sensors that have variable sensitivity that I want to be super sensitive before they are tripped... then to conserve battery i'd want to drop the sensitivity while operating under constant movement and then return to super sensitive after a period of inactivity. Does that make sense? I'll be attaching them to the oil boiler, sump pump, washer, dryer and the dishwasher in case you were curious what I was doing w/ them. This use case is allows me to fire automations that allow me to maximize battery life of the sensors.